pymonetdb/filetransfer.py

```diff
@@ -0,0 +1,621 @@
+
+        normalized = data.replace(b"\r\n", b"\n")
+
+        if normalized[-1] == 13:  # \r
+            # not sure if it will be followed by \n, move it to pending
+            self.pending = True
+            normalized = memoryview(normalized)[:-1]
+        else:
+            self.pending = False
+
+        n = self.inner.write(normalized)
+        assert n == len(normalized)
+        return len(data)
+
+    def flush(self):
+        return self.inner.flush()
+
+    def close(self):
+        if self.pending:
+            self.inner.write(b"\r")
+            self.pending = False
+        return self.inner.close()
+
+
+class SafeDirectoryHandler(Uploader, Downloader):
+    """
+    File transfer handler which uploads and downloads files from a given
+    directory, taking care not to allow access to files outside that directory.
+    Instances of this class can be registered using the pymonetb.Connection's
+    set_uploader() and set_downloader() methods.
+
+    When downloading text files, the downloaded text is converted according to
+    the `encoding` and `newline` parameters, if present. Valid values for
+    `encoding` are any encoding known to Python, or None. Valid values for
+    `newline` are `"\\\\n"`, `"\\\\r\\\\n"` or None. None means to use the
+    system default.
+
+    For binary up- and downloads, no conversions are applied.
+
+    When uploading text files, the `encoding` parameter indicates how the text
+    is read and `newline` is mostly ignored: both `\\\\n` and `\\\\r\\\\n` are
+    valid line endings. The exception is that because the server expects its
+    input to be `\\\\n`-terminated UTF-8 text,  if you set encoding to "utf-8"
+    and newline to "\\\\n", text mode transfers are performed as binary, which
+    improves performance. For uploads, only do this if you are absolutely,
+    positively sure that all files in the directory are actually valid UTF-8
+    encoded and have Unix line endings.
+
+    If `compression` is set to True, which is the default, the
+    SafeDirectoryHandler will automatically compress and decompress files with
+    extensions .gz, .bz2, .xz and .lz4. Note that the first three algorithms are
+    built into Python, but LZ4 only works if the lz4.frame module is available.
+    """
+
+    def __init__(self, dir, encoding: str = None, newline=None, compression=True):
+        self.dir = Path(dir).resolve()
+        self.encoding = encoding
+        self.is_utf8 = (self.encoding and (codecs.lookup('utf-8') == codecs.lookup(self.encoding)))
+        self.newline = newline
+        self.compression = compression
+
+    def secure_resolve(self, filename) -> Optional[Path]:
+        p = self.dir.joinpath(filename).resolve()
+        if str(p).startswith(str(self.dir)):
+            return p
+        else:
+            return None
+
+    def handle_upload(self, upload: Upload, filename: str, text_mode: bool, skip_amount: int):
+        """:meta private:"""
```

Comment:
?

Reply:
I wanted SafeDirectoryHandler to appear in the API documentation but without this annotation it will also add documentation for this overridden method, which has already been documented when it was defined on Uploader. Suppressing it makes the documentation clearer and more to the point. 
I'll add a comment to clarify this.